lib/plausible/stats/sql/special_metrics.ex

```diff
@@ -12,73 +12,114 @@ defmodule Plausible.Stats.SQL.SpecialMetrics do
   import Plausible.Stats.Util
 
   def add(q, site, query) do
-    q
-    |> maybe_add_percentage_metric(site, query)
-    |> maybe_add_global_conversion_rate(site, query)
-    |> maybe_add_group_conversion_rate(site, query)
-    |> maybe_add_scroll_depth(query)
+    Enum.reduce(query.metrics, q, fn
+      :exit_rate, q -> add_exit_rate(q, query, site)
+      :percentage, q -> add_percentage_metric(q, site, query)
+      :conversion_rate, q -> add_global_conversion_rate(q, site, query)
+      :group_conversion_rate, q -> add_group_conversion_rate(q, site, query)
+      :scroll_depth, q -> add_scroll_depth(q, query)
+      _, q -> q
+    end)
   end
 
-  defp maybe_add_percentage_metric(q, site, query) do
-    if :percentage in query.metrics do
-      total_query =
-        query
-        |> remove_filters_ignored_in_totals_query()
-        |> Query.set(
-          dimensions: [],
-          include_imported: query.include_imported,
-          pagination: nil
-        )
+  defp add_exit_rate(q, query, site) do
+    total_pageviews_query =
+      query
+      |> Query.remove_top_level_filters(["visit:exit_page"])
+      |> remove_filters_ignored_in_totals_query()
+      |> Query.set(
+        pagination: nil,
+        order_by: [],
+        metrics: [:pageviews],
+        include_imported: query.include_imported,
+        dimensions: ["event:page"]
+      )
 
+    joined_q =
       q
-      |> select_merge_as([], total_visitors_subquery(site, total_query, query.include_imported))
-      |> select_merge_as([], %{
-        percentage:
+      |> join(:left, [..., s], p in subquery(SQL.QueryBuilder.build(total_pageviews_query, site)),
+        on:
+          selected_as(^shortname(query, "visit:exit_page")) ==
+            field(p, ^shortname(total_pageviews_query, "event:page"))
+      )
+
+    if query.include_imported do
+      joined_q
+      |> select_merge_as([_s, _i, p], %{
```

Comment:
How would you feel about

```suggestion
      |> select_merge_as([..., p], %{
```

IMO slightly less fragile to other query building changes outside of this module causing unexpected breakage.

(See also below)

Reply:
Agreed! https://github.com/plausible/analytics/pull/5289/commits/e544d4f2d37c09c6c1a987ea762fa8deab319a12